src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/ProjectSystem/RazorProjectInfoEndpointPublisher.cs

```diff
@@ -0,0 +1,205 @@
+    // so that we don't publish half-finished projects, which can cause things like Semantic coloring to "flash"
+    // when they update repeatedly as they load.
+    private static bool ProjectWorkspacePublishable(IProjectSnapshot? project)
+    {
+        return project?.ProjectWorkspaceState != null;
+    }
+
+    private void EnqueuePublish(IProjectSnapshot projectSnapshot)
+    {
+        _workQueue.AddWork((Project: projectSnapshot, Removal: false));
+    }
+
+    private void EnqueueRemoval(IProjectSnapshot projectSnapshot)
+    {
+        _workQueue.AddWork((Project: projectSnapshot, Removal: true));
+    }
+
+    private ValueTask ProcessBatchAsync(ImmutableArray<(IProjectSnapshot Project, bool Removal)> workItems, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in workItems.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return default;
+            }
+
+            if (workItem.Removal)
+            {
+                RemovePublishingData(workItem.Project, cancellationToken);
+            }
+            else
+            {
+                ImmediatePublish(workItem.Project, cancellationToken);
+            }
+        }
+
+        return default;
+    }
+
+    private void RemovePublishingData(IProjectSnapshot projectSnapshot, CancellationToken cancellationToken)
+    {
+        // This should never get called if we are inactive, so don't check _active flag
+        ImmediatePublish(projectSnapshot.Key, encodedProjectInfo: null, cancellationToken);
+    }
+
+    private void ImmediatePublish(IProjectSnapshot projectSnapshot, CancellationToken cancellationToken)
+    {
+        if (!_active)
+        {
+            return;
+        }
+
+        using var stream = new MemoryStream();
+
+        var projectInfo = projectSnapshot.ToRazorProjectInfo(projectSnapshot.IntermediateOutputPath);
+        projectInfo.SerializeTo(stream);
+        var base64ProjectInfo = Convert.ToBase64String(stream.ToArray());
+
+        ImmediatePublish(projectSnapshot.Key, base64ProjectInfo, cancellationToken);
+    }
+
+    private void ImmediatePublish(ProjectKey projectKey, string? encodedProjectInfo, CancellationToken cancellationToken)
+    {
+        // This might be getting called after getting dequeued from work queue,
+        // check if the work got cancelled before doing anything.
+        if (cancellationToken.IsCancellationRequested)
+        {
+            return;
+        }
+
+        var parameter = new ProjectInfoParams()
+        {
+            ProjectKeyId = projectKey.Id,
+            ProjectInfo = encodedProjectInfo
+        };
+
+        _ = _requestInvoker.ReinvokeRequestOnServerAsync<ProjectInfoParams, object>(
+                LanguageServerConstants.RazorProjectInfoEndpoint,
+                RazorLSPConstants.RazorLanguageServerName,
+                parameter,
+                cancellationToken);
```

Comment:
Please use the `Forget()` extension method rather than assigning a task to a discard. That helps us used Find All References to spot these in the future if we want to clean them up.